### PR TITLE
[READY] SSdiscord and Round Notifications

### PR DESCRIPTION
**Pull request by AffectedArc07**
## About The Pull Request
**NOTE** I am still new to how SQL changes work here, so please tell me if I fucked anything up

This PR adds multiple things, and will be explained below

**SSdiscord**
SSdiscord will (On Initialize) send a message out to the discord, using the data from `notify.json` (Explained further down), to ping the requested users. Discord links are synced between servers, as they are on the DB. Notify status will __NOT__ sync between servers as its local. This is intentional because if you notify on Sybil, Terry restarting would trigger it.

Every 5 minutes, SSdiscord will fire, and will log a file of those who want to be notified to the disk. This is done on `fire()` instead of only `Shutdown()`, so notifications still happen if the server is hard-restarted. The file is still force-written on `Shutdown()` anyway, so the file is always up to date on roundend.

**Linking Accounts**
To use the notify system, you first link your ckey to your discord ID, using the verb `Link Discord Account`. If a user is unaware of how to obtain their ID, a link to a guide on how to get the ID, provided by Discord Support. After pressing this, you will be instructed on discord to do `!tgs verify (ckey)`. After this, you will be able to use the notify ingame.

If you were to run this verb again, it will tell you your account is already linked, and will ask you if you wish to link to a different account, or cancel. If you request to relink, it will do the verification process again.

**Admin Verb**
Admins now have a verb that can lookup a discord ID by ckey, and vice versa, with an option to unlink for whatever reason

**!tgs notify**
This PR makes `!tgs notify` command now works, is available to all players instead of admins, and fires off at roundstart, instead of roundend. This command works whether you have linked your account or not, but the ingame command only works if you have linked your account. Both of these are in sync so if you notify ingame, running `!tgs notify` will un-notify you.

**!tgs verify**
This command is purely used to verify ckeys to IDs, to avoid false-linking from occuring.
 

## Why It's Good For The Game
A notify system is nice to have so that people can be told when a round is going up, and that is useful to have when you have 3-4 servers up and wish to know when one specific one is up.

## Changelog
:cl: AffectedArc07
add: You can now link your ingame account to your discord account
add: You can now set notify status ingame, and be notified on discord when a round restarts
admin: There is now an admin panel to lookup IDs to ckeys, and vice versa
/:cl:

*Yes I know its shitcode, but its my shitcode so I have to be proud*

### CODE REVIEW

**Comment by AffectedArc07** [comment]
Dont know if my last CL line is necessary, but its there anyway

**Comment by optimumtact** [comment]
it isn't, remove it

**Comment by AffectedArc07** [comment]
Done

**Comment by nfreader** [comment]
I feel like `discord_id` should go on the `player` table instead of making a brand new table to keep track of.

**Comment by francinum** [comment]
There should be a way to actually ***authenticate*** the user to their tag, instead of blindly trusting user input.

**Comment by AffectedArc07** [comment]
V2 has been released:
- It now uses the players table
- It is now `SSdiscord` and has some other procs thrown into it
- There is now secure 2-way linking of the accounts

This is not yet fully tested, because I need to install TGS on my laptop. Will report back when I have

**Comment by AffectedArc07** [comment]
I installed TGS, but even after relaunching the server and recompiling __multiple__ times, my server would not detect it was running on TGS (Im running 3.2.6.2). If anyone has a working TGS setup or support for this, I would greatly appreciate it

File: code/controllers/subsystem/discord.dm [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+2] On fire, it will write that to the disk, as long as conditions above are correct
+
+END ROUND:
+1] The file is force-saved, incase it hasnt fired at end round
+
+This is an absolute clusterfuck, but its my clusterfuck -aa07
+*/
+
+SUBSYSTEM_DEF(discord)
+	name = "DIscord"
```
Comment (AffectedArc07) [inline code/controllers/subsystem/discord.dm]: Yes I know this is badly capitalised, I will fix this when I have to make the fix PRs for the inevitable other issues in this

**Comment by optimumtact** [comment]
>If anyone has a working TGS setup or support for this, I would greatly appreciate it

**Comment by optimumtact** [comment]
@MrStonedOne for sql changes

**Comment by daklj** [comment]
who cares, only kids use discord

**Comment by MrStonedOne** [comment]
```
!tgs notify
This PR makes !tgs notify command now works, is available to all players instead of admins, and fires off at roundstart, instead of roundend.
```

This was round end because server operators use it to know when the round has ended for things that require taking the server down. Its a minor change but if say, graceful stop had be used to have the server shutdown after the current round, the notify would fail to fire.

This can be fixed in a later pr though, since this pr has already sat around long enough.

File: code/modules/discord/accountlink.dm [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+	var/stored_id = SSdiscord.lookup_id(usr.ckey)
+	if(!stored_id) // Account is not linked
+		var/know_how = alert("Do you know how to get a discord user ID? This ID is NOT your discord username and numbers! (Pressing NO will open a guide)","Question","Yes","No")
+		if(know_how == "No") // Opens discord support on how to collect IDs
+			src << link("https://support.discordapp.com/hc/en-us/articles/206346498-Where-can-I-find-my-User-Server-Message-ID")
+
+		var/entered_id = input("Please enter your Discord ID (17 digits)", "Enter Discord ID", null, null) as text|null
+		SSdiscord.account_link_cache[lowertext(usr.ckey)] = "[entered_id]" // Prepares for TGS-side verification
+		alert(usr, "Account link started. Please do \"!tgs verify [lowertext(usr.ckey)]\" in the Discord to successfuly verify your account")
```
Comment (MrStonedOne) [inline code/modules/discord/accountlink.dm]: Minor note, on codebases with multiple servers (/tg/, and, well, /tg/) running `!tgs verify` will trigger all of the bots at once, and they would really need to run `@botname verify usr.ckey` Also, ckey is already lower text, key is the one that has cases.

**Review by MrsTonedOne — APPROVED** [review]
Sql changes commited.

**Comment by AffectedArc07** [comment]
I'll fix these when I get round to it, and I also managed to make TGS half work now, so ill also test that

**Comment by AffectedArc07** [comment]
DNM, and surely theres a way to test code on TGS without committing each time

File: code/controllers/subsystem/discord.dm [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+
+// Returns ID from ckey
+/datum/controller/subsystem/discord/proc/lookup_id(lookup_ckey)
+	var/stored_id
+	var/datum/DBQuery/get_discord_id = SSdbcore.NewQuery("SELECT discord_id FROM [format_table_name("player")] WHERE ckey = '[sanitizeSQL(lookup_ckey)]'")
+	if(get_discord_id.Execute())
+		while(get_discord_id.NextRow())
+			stored_id = get_discord_id.item[1]
+	qdel(get_discord_id)
+	return stored_id
```
Comment (Jordie0608) [inline code/controllers/subsystem/discord.dm]: Restructure this as:
```
/datum/controller/subsystem/discord/proc/lookup_id(lookup_ckey)
	var/datum/DBQuery/query_get_discord_id = SSdbcore.NewQuery("SELECT discord_id FROM [format_table_name("player")] WHERE ckey = '[sanitizeSQL(lookup_ckey)]'")
	if(!query_get_discord_id.Execute())
		qdel(query_get_discord_id)
		return
	if(query_get_discord_id.NextRow())
		. = query_get_discord_id.item[1]
	qdel(query_get_discord_id)
```

File: code/controllers/subsystem/discord.dm [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+// Returns ckey from ID
+/datum/controller/subsystem/discord/proc/lookup_ckey(lookup_id)
```
Comment (Jordie0608) [inline code/controllers/subsystem/discord.dm]: Same for this one.

File: code/modules/discord/accountlink.dm [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+		to_chat(src, "<span class='notice'>The server is still starting up. Please wait before attempting to link your account </span>")
+		return
+
+	if(!SSdiscord.enabled)
+		to_chat(src, "<span class='warning'>This feature requires the server is running on the TGS toolkit, </span>")
+		return
+
+	var/stored_id = SSdiscord.lookup_id(usr.ckey)
+	if(!stored_id) // Account is not linked
+		var/know_how = alert("Do you know how to get a discord user ID? This ID is NOT your discord username and numbers! (Pressing NO will open a guide)","Question","Yes","No")
```
Comment (Jordie0608) [inline code/modules/discord/accountlink.dm]: Can you add a cancel button for this.

File: code/modules/discord/accountlink.dm [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+		if(know_how == "No") // Opens discord support on how to collect IDs
+			src << link("https://support.discordapp.com/hc/en-us/articles/206346498-Where-can-I-find-my-User-Server-Message-ID")
+
+		var/entered_id = input("Please enter your Discord ID (17 digits)", "Enter Discord ID", null, null) as text|null
+		SSdiscord.account_link_cache[lowertext(usr.ckey)] = "[entered_id]" // Prepares for TGS-side verification
+		alert(usr, "Account link started. Please ping the bot of the server you\'re currently on, follows by \"verify [usr.ckey]\" in the Discord to successfuly verify your account (Example: @Mr_Terry notify [usr.ckey]")
+
+	else // Account is already linked
+		var/choice = alert("You already have the Discord Account [stored_id] linked to [usr.ckey]. Would you like to link a different account?","Already Linked","Yes","No")
```
Comment (Jordie0608) [inline code/modules/discord/accountlink.dm]: And this.

File: code/modules/discord/accountlink.dm [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+	var/stored_id = SSdiscord.lookup_id(usr.ckey)
+	if(!stored_id) // Account is not linked
+		var/know_how = alert("Do you know how to get a discord user ID? This ID is NOT your discord username and numbers! (Pressing NO will open a guide)","Question","Yes","No")
+		if(know_how == "No") // Opens discord support on how to collect IDs
+			src << link("https://support.discordapp.com/hc/en-us/articles/206346498-Where-can-I-find-my-User-Server-Message-ID")
+
+		var/entered_id = input("Please enter your Discord ID (17 digits)", "Enter Discord ID", null, null) as text|null
+		SSdiscord.account_link_cache[lowertext(usr.ckey)] = "[entered_id]" // Prepares for TGS-side verification
+		alert(usr, "Account link started. Please ping the bot of the server you\'re currently on, follows by \"verify [usr.ckey]\" in the Discord to successfuly verify your account (Example: @Mr_Terry notify [usr.ckey]")
```
Comment (Jordie0608) [inline code/modules/discord/accountlink.dm]: `on, follows by`
Should be followed by no?

Comment (Jordie0608) [reply code/modules/discord/accountlink.dm]: Also you say to use verify, but the example is with notify.

File: code/modules/discord/accountlink.dm [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+	else // Account is already linked
+		var/choice = alert("You already have the Discord Account [stored_id] linked to [usr.ckey]. Would you like to link a different account?","Already Linked","Yes","No")
+		if(choice == "Yes")
+			var/know_how = alert("Do you know how to get a discord user ID? This ID is NOT your discord username and numbers! (Pressing NO will open a guide)","Question","Yes","No")
```
Comment (Jordie0608) [inline code/modules/discord/accountlink.dm]: Cancel here too.

File: code/modules/discord/manipulation.dm [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+		return
+
+	if(!SSdiscord.enabled)
+		to_chat(usr, "<span class='warning'>TGS is not enabled</span>")
+		return
+
+	var/lookup_choice = alert(usr, "Do you wish to lookup account by ID or ckey?", "Lookup Type", "ID", "Ckey", "Cancel")
+	switch(lookup_choice)
+		if("ID")
+			var/lookup_id = input(usr,"Enter Discord ID to lookup ckey") as text
```
Comment (Jordie0608) [inline code/modules/discord/manipulation.dm]: Use text|null for the cancel button.

File: code/modules/discord/manipulation.dm [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+			var/returned_ckey = SSdiscord.lookup_id(lookup_id)
+			if(returned_ckey)
+				var/unlink_choice = alert(usr, "Discord ID [lookup_id] is linked to Ckey [returned_ckey]. Do you wish to unlink or cancel?", "Account Found", "Unlink", "Cancel")
+				if(unlink_choice == "Unlink")
+					SSdiscord.unlink_account(returned_ckey)
+			else
+				to_chat(usr, "<span class='warning'>Discord ID <b>[lookup_id]</b> has no associated ckey</span>")
+		if("Ckey")
+			var/lookup_ckey = input(usr,"Enter Ckey to lookup ID") as text
```
Comment (Jordie0608) [inline code/modules/discord/manipulation.dm]: Same.

**Review by Jordie0608 — CHANGES_REQUESTED** [review]

File: code/modules/discord/accountlink.dm (hunk 6 above)
Comment (AffectedArc07) [reply code/modules/discord/accountlink.dm]: `No` is the cancel here

**Comment by AffectedArc07** [comment]
I just realised I can use TGS static files to change this without committing

God I am retarded

**Comment by AffectedArc07** [comment]
image
And now it broke

**Comment by AffectedArc07** [comment]
This took so long to do it wasnt even funny. The stuff works now atleast

**Comment by AffectedArc07** [comment]
@MrStonedOne want me to tack in `!tgs endnotify` to trigger again at end round to get the feature so that server operators can do what they need?

**Comment by optimumtact** [comment]
mso never responds just add it

**Comment by SpaceManiac** [comment]
@AffectedArc07 Discord pings are being sent to IRC as well, is this something you can fix or a TGS issue?

**Comment by AffectedArc07** [comment]
@SpaceManiac uhhhhhhhh I dont know. I aint fully familiar with TGS. I wrote the system around discord because I didnt realise anyone even used IRC anymore. I dont think TGS has a way of sending specific messages out to each seperate platform.

**Pull request merged**